sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/interop/access/LLVMInteropType.java

```diff
@@ -313,27 +391,124 @@ public Method findMethodByArgumentsWithSelf(String memberName, Object[] argument
-            Object[] newArgs = new Object[arguments.length + 1];
-            newArgs[0] = receiver;
-            for (int i = 1; i < arguments.length; i++) {
-                newArgs[i] = arguments[i - 1];
+    }
+
+    public static final class VTable {
+        final HashMap<Long, Method> table;
+        final Clazz clazz;
+
+        VTable(Clazz clazz) {
+            this.clazz = clazz;
+            this.table = new HashMap<>();
+            List<Clazz> list = new LinkedList<>(clazz.getSuperClasses());
+            list.add(0, clazz);
+            do {
+                Clazz c = list.remove(0);
+                list.addAll(c.superclasses);
+                for (Method m : c.methods) {
+                    if (m != null && m.getVirtualIndex() >= 0) {
+                        table.putIfAbsent(m.virtualIndex, m);
+                    }
+                }
+            } while (list.size() > 0);
+        }
+
+        public LLVMInteropType.Method findMethod(long virtualIndex) {
+            final LLVMInteropType.Method m = table.get(virtualIndex);
+            if (m == null) {
+                CompilerDirectives.transferToInterpreter();
+                throw new NoSuchElementException(String.format("No method in %s with virtualIndex %d", clazz.name, virtualIndex));
             }
-            Method method = findMethodByArgumentsWithSelf(memberName, newArgs);
-            if (method == null) {
-                throw UnknownIdentifierException.create(memberName);
+            return m;
+        }
+    }
+
+    @ExportLibrary(value = InteropLibrary.class)
+    public static final class VTableObjectPair implements TruffleObject {
+        private final VTable vtable;
+        private final Object foreign;
+
+        private VTableObjectPair(VTable vtable, Object foreign) {
+            this.vtable = vtable;
+            this.foreign = foreign;
+        }
+
+        public static VTableObjectPair create(VTable vtable, Object foreign) {
+            return new VTableObjectPair(vtable, foreign);
+        }
+
+        @SuppressWarnings("static-method")
+        @ExportMessage
+        boolean hasArrayElements() {
+            return true;
+        }
+
+        @ExportMessage
+        Object readArrayElement(long index) throws UnsupportedMessageException {
+            try {
+                final String methodName = vtable.findMethod(index).getName();
+                try {
+                    Object readMember = InteropLibrary.getUncached(foreign).readMember(foreign, methodName);
+                    return new RemoveSelfArgument(readMember);
+                } catch (UnknownIdentifierException e) {
+                    final String msg = String.format("External method %s (identifier \"%s\") not found in type %s", methodName, e.getUnknownIdentifier(), foreign.getClass().getSimpleName());
+                    throw new IllegalStateException(msg);
```

Comment:
Here, I am not sure whether this solution is correct, because from an interop view point, this is an UnknownIdentifierException (=i.e. the foreign object, e.g. JavaScript, does not have a specific field/method). However, from a C++ side, it is an array access by index.